pkg/health/membergarbagecollector/membergarbagecollector.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package membergarbagecollector
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gardener/etcd-backup-restore/pkg/errors"
+	"github.com/gardener/etcd-backup-restore/pkg/etcdutil"
+	"github.com/gardener/etcd-backup-restore/pkg/miscellaneous"
+	utils "github.com/gardener/etcd-backup-restore/pkg/snapstore"
+	brtypes "github.com/gardener/etcd-backup-restore/pkg/types"
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	etcdclient "github.com/gardener/etcd-backup-restore/pkg/etcdutil/client"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	podNamespace = "POD_NAMESPACE"
+	podName      = "POD_NAME"
+)
+
+// MGCChecker is an interface used to unit test functions in this package
+type MGCChecker interface {
+	RemoveSuperfluousMembers(ctx context.Context) error
+}
+
+// MemberGarbageCollector contains information to regularly cleanup superfluous ETCD members
+type MemberGarbageCollector struct {
+	logger       *logrus.Entry
+	gcTimer      *time.Timer
+	k8sClient    client.Client
+	podNamespace string
+	etcdName     string
+}
+
+func init() {
+	druidv1alpha1.AddToScheme(scheme.Scheme)
+}
+
+// NewMemberGarbageCollector returns the member garbage collect object
+func NewMemberGarbageCollector(logger *logrus.Entry, clientSet client.Client) (*MemberGarbageCollector, error) {
+	if clientSet == nil {
+		return nil, &errors.EtcdError{
+			Message: "nil kubernetes clientset passed, can not create etcd member garbage collector",
+		}
+	}
+	namespace, err := utils.GetEnvVarOrError(podNamespace)
+	if err != nil {
+		logger.Fatalf("POD_NAMESPACE env var not present: %v", err)
```

Comment:
imo, better to fail early rather than continue with incorrect configuration

Reply:
>imo, better to fail early rather than continue with incorrect configuration

I would say it depends on what priorities we set, in other words .... Do `etcd-member-gc` configurations are that important that we are ready to cause outage as backup-restore crash will also fails the readinessProbe?

Reply:
I would imagine since `etcd-member-gc` can anyway be turned on or off using its flag, than it would be fine to crash backup-restore for a configuration issue
I however do not have a hard opinion on this, so if you feel this is a bit too strict, I could update this

Reply:
hmm, so Let's keep it the `Fatalf` as [pkg/health/heartbeat/heartbeat.go](https://github.com/gardener/etcd-backup-restore/blob/322d3c74fcd5ea7be9961ce5e9d7ffed48ecc34e/pkg/health/heartbeat/heartbeat.go#L63-L69) is also throwing the `Fatalf` error